src/Columns/ColumnSparse.cpp

```diff
@@ -0,0 +1,762 @@
+    if (offsets->size() + 1 != values->size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Values size ({}) is inconsistent with offsets size ({})", values->size(), offsets->size());
+
+    if (_size < offsets->size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Size of sparse column ({}) cannot be lower than number of non-default values ({})", _size, offsets->size());
+
+#ifndef NDEBUG
+    const auto & offsets_data = getOffsetsData();
+    const auto * it = std::adjacent_find(offsets_data.begin(), offsets_data.end(), std::greater_equal<UInt64>());
+    if (it != offsets_data.end())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Offsets of ColumnSparse must be strictly sorted");
+#endif
+}
+
+MutableColumnPtr ColumnSparse::cloneResized(size_t new_size) const
+{
+    if (new_size == 0)
+        return ColumnSparse::create(values->cloneEmpty());
+
+    if (new_size >= _size)
+        return ColumnSparse::create(IColumn::mutate(values), IColumn::mutate(offsets), new_size);
+
+    auto res = ColumnSparse::create(values->cloneEmpty());
+    res->insertRangeFrom(*this, 0, new_size);
+    return res;
+}
+
+bool ColumnSparse::isDefaultAt(size_t n) const
+{
+    return getValueIndex(n) == 0;
+}
+
+bool ColumnSparse::isNullAt(size_t n) const
+{
+    return values->isNullAt(getValueIndex(n));
+}
+
+Field ColumnSparse::operator[](size_t n) const
+{
+    return (*values)[getValueIndex(n)];
+}
+
+void ColumnSparse::get(size_t n, Field & res) const
+{
+    values->get(getValueIndex(n), res);
+}
+
+bool ColumnSparse::getBool(size_t n) const
+{
+    return values->getBool(getValueIndex(n));
+}
+
+Float64 ColumnSparse::getFloat64(size_t n) const
+{
+    return values->getFloat64(getValueIndex(n));
+}
+
+Float32 ColumnSparse::getFloat32(size_t n) const
+{
+    return values->getFloat32(getValueIndex(n));
+}
+
+UInt64 ColumnSparse::getUInt(size_t n) const
+{
+    return values->getUInt(getValueIndex(n));
+}
+
+Int64 ColumnSparse::getInt(size_t n) const
+{
+    return values->getInt(getValueIndex(n));
+}
+
+UInt64 ColumnSparse::get64(size_t n) const
+{
+    return values->get64(getValueIndex(n));
+}
+
+StringRef ColumnSparse::getDataAt(size_t n) const
+{
+    return values->getDataAt(getValueIndex(n));
+}
+
+ColumnPtr ColumnSparse::convertToFullColumnIfSparse() const
+{
+    return values->createWithOffsets(getOffsetsData(), (*values)[0], _size, /*shift=*/ 1);
+}
+
+void ColumnSparse::insertData(const char * pos, size_t length)
+{
+    _size += length;
+    return values->insertData(pos, length);
+}
+
+/// TODO: maybe need to reimplement it.
+StringRef ColumnSparse::serializeValueIntoArena(size_t n, Arena & arena, char const *& begin) const
+{
+    return values->serializeValueIntoArena(getValueIndex(n), arena, begin);
+}
+
+const char * ColumnSparse::deserializeAndInsertFromArena(const char * pos)
+{
+    ++_size;
+    return values->deserializeAndInsertFromArena(pos);
```

Comment:
Also I would check `isDefaultAt` here, or we will loose sparsity after serialize/deserialize